### DISCUSSION TITLE
Storybook using web-components broken after upgrade to 7.6.4

### DISCUSSION THREAD

**Discussion by ccsenter** [storybookjs/storybook#25181] [Help] [answered]
### Summary

We've been using 7.0.7 for a while and decided to catch up with latest updates so using storybook auto-migrate/update I updated migrated this web-components project from 7.0.7 to 7.6.4
After the upgrade, the manager UI renders successfully I can see the stories in the navigator however the preview frame shows an error (see the attached pic). 
The build also shows some warnings in the console which I guess is related to the error from the UI.
The console shows this for every .mdx page I have in the project:
```
WARN export 'Basic' (imported as 'Stories') was not found in './button.stories.ts' (possible exports: __namedExportsOrder)
WARN export 'Basic' (imported as 'Stories') was not found in './button.stories.ts' (possible exports: __namedExportsOrder)
```
In the project I have *.mdx files that are importing stories from am "associated" .stories.tx file. Something like:
```
import { ArgTypes, Canvas, Meta, Story } from '@storybook/blocks';
import 'oj-c/button';
import * as Stories from './button.stories.ts';

<Meta of={Stories} />

# Button

The `oj-c-button` component is a push button.

## Usage

Here's how to implement a basic `oj-c-button`.

<Canvas of={Stories.Basic}>
  <Story of={Stories.Basic} height={350} />
</Canvas>
```
Of course the .stories.ts file has a default export and a named export.
![error](https://github.com/storybookjs/storybook/assets/91220183/636cfef8-f40e-4dcf-a003-58adbc7883d1)

**Comment by ccsenter** [comment]
One additional fact that might be interesting is that if I create a static build and run SB after that, the error is a bit different in the preview. This is what I'm seeing:
```
Cannot access '__namedExportsOrder' before initialization
ReferenceError: Cannot access '__namedExportsOrder' before initialization
    at Module.__namedExportsOrder (http://ccsenter-dev.subnet3ad1phx.devdevcsphx.oraclevcn.com/jetdev/ccsenter-sb-upgrade-OJC-DEV/storybook-static/components-oj-c-button-__dev__-button-doc-mdx.37ada1fd.iframe.bundle.js:2:1661)
    at processCSFFile (http://ccsenter-dev.subnet3ad1phx.devdevcsphx.oraclevcn.com/jetdev/ccsenter-sb-upgrade-OJC-DEV/storybook-static/sb-preview/runtime.js:41:91)
    at StoryStore.memoizerific [as processCSFFileWithCache] (http://ccsenter-dev.subnet3ad1phx.devdevcsphx.oraclevcn.com/jetdev/ccsenter-sb-upgrade-OJC-DEV/storybook-static/sb-preview/runtime.js:1:4751)
    at http://ccsenter-dev.subnet3ad1phx.devdevcsphx.oraclevcn.com/jetdev/ccsenter-sb-upgrade-OJC-DEV/storybook-static/sb-preview/runtime.js:47:8515
    at async Promise.all (index 1)
    at async StoryStore.loadEntry (http://ccsenter-dev.subnet3ad1phx.devdevcsphx.oraclevcn.com/jetdev/ccsenter-sb-upgrade-OJC-DEV/storybook-static/sb-preview/runtime.js:47:11053)
    at async MdxDocsRender.prepare (http://ccsenter-dev.subnet3ad1phx.devdevcsphx.oraclevcn.com/jetdev/ccsenter-sb-upgrade-OJC-DEV/storybook-static/sb-preview/runtime.js:92:2746)
    at async PreviewWeb.renderSelection (http://ccsenter-dev.subnet3ad1phx.devdevcsphx.oraclevcn.com/jetdev/ccsenter-sb-upgrade-OJC-DEV/storybook-static/sb-preview/runtime.js:101:3102)
```

**Comment by valentinpalkovic** [comment]
Can you create a minimal reproduction?

**Reply by ccsenter** [reply]
Will be hard if possible at all. We use ts-loader in the webpackFinal (in main.js) entry and we are using some custom transformers with tsc. It is a fairly complicated setup. 
Maybe you can help me understand where to look for the issue. So by default I think babel-loader is used with webpack to transpile the stories files. And I think this loader is also using the ``babel-plugin-named-exports-order`` plugin to handle the named and default exports from a story file. In my case since I'm using ts-loader and not babel-loader, there might be a gap in the functionality. Wdyt?
Comparing with another SB project but where we don't use web-components and the loader is babel-loader I see in the transpiled, webpack generated stories bundle files declarations like:
```const __namedExportsOrder=["Default"] ``` whereas in the problematic case it's just:
```const __namesExportOrder=[]```

**Reply by valentinpalkovic** [reply]
In Storybook 7.6.4 we don't use babel-plugin-named-exports-order anymore, but instead we have rewritten it to use a webpack loader instead to support both, Babel and SWC out of the box. Have you tried switching to swc instead?

**Reply by ccsenter** [reply]
I see. Now that I looked more into the SB code I see that in export-order-loader.ts (builder-webpack5). About SWC...I've never used in the past. Are you suggesting to use webpack with SWC (using swc-loader) or completely replace webpack? How would that work since SB only supports vite and webpack as builder/bundler?

**Reply by ccsenter** [reply]
I was doing some debugging in the export-order-loader and the problem I see (and not sure where this stems from) is that at line https://github.com/storybookjs/storybook/blob/b4acf3a61734d4dc5d054dd2864a186e03f12734/code/builders/builder-webpack5/src/loaders/export-order-loader.ts#L11
once the source is parsed it does not come back with any exports. Doing a comparison between the two projects we have (one using web-components the other is a Preact based project) with regards to the line above, in the preact project the source is still the TS source so there are imports/exports whereas in the web-components based project, the source received by the loader is already a transpiled source, there are no imports/exports.
I'm scratching my head for the cause of this difference.

**Reply by valentinpalkovic** [reply]
Could you show me the content of your `.storybook/main.ts` file?

So in [Storybook 7.6](https://storybook.js.org/blog/storybook-7-6/) we have introduced SWC support. The Webpack5 traditionally used babel under the hood to transpile the files. Now you can switch over to swc instead. You just have to set a flag in your `.storybook/main.ts` file:

```ts
// .storybook/main.ts

export default {
  framework: {
    name: '@storybook/your-framework', // react-webpack5|next|vue-webpack5|etc
    options: { builder: { useSWC: true } }
  }
};
```

My guess is, that your ts-loader already transforms the file, before the export-order-loader plugin can process it. Could you also send me the content of your transpiled file, which doesn't have any export statements anymore?

**Reply by ccsenter** [reply]
Hi Valentin, much appreciated your help and suggestions. Using SWC did not change the outcome, same warnings in the terminal about 
```
export 'Basic' (imported as 'Stories') was not found in './button.stories.ts' (possible exports: __namedExportsOrder)
```
I attached ``.storybok/main.ts`` but I have to underline that it does contain some complex configuration that
[main.txt](https://github.com/storybookjs/storybook/files/13668167/main.txt)
 was needed for our custom elements

**Reply by ccsenter** [reply]
Forgot to mention that if I remove the webpack rule to use the ``export-order-loader.js`` loader, then the build finishes and SB renders. I also attached a transpiled file that gets to be processed by ``export-order-loader.js``. 
[transpiled.txt](https://github.com/storybookjs/storybook/files/13668660/transpiled.txt)

**Reply by valentinpalkovic** [reply]
I think the issue is that your ts-loader outputs CJS code instead of ESM. The export statements are already transformed. Can you force your ts-loader not to transform any export/import statements? I think that's totally okay, since Webpack will use its own module resolution system, which works just okay with ESM exports/imports.

**Reply by ccsenter** [reply, marked as answer]
That's what I think too. Now that I looked at the tsconfig provided for the ts-loader the module is actually 'amd'. After rethinking what is going on I think I managed to overcome this problem, by explicitly excluding stories files from the ts-loader compilation process.
Now the stories are rendering fine!
Thanks for all your help Valentin!
